src/main/scala/com/databricks/labs/overwatch/MultiWorkspaceDeployment.scala

```diff
@@ -0,0 +1,397 @@
+package com.databricks.labs.overwatch
+
+import com.databricks.labs.overwatch.pipeline._
+import com.databricks.labs.overwatch.utils._
+import com.databricks.labs.overwatch.validation.{ConfigColumns, DeploymentValidation}
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.sql.functions.lit
+import org.apache.spark.sql.{DataFrame, Dataset, Row}
+
+import java.text.SimpleDateFormat
+import java.time.LocalDateTime
+import java.util.Date
+import java.util.concurrent.Executors
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
+
+/** *
+ * MultiWorkspaceDeployment class is the main class which runs the deployment for multiple workspaces.
+ *
+ * @params configCsvPath: path of the csv file which will contain the different configs for the workspaces.
+ * @params tempOutputPath: location which will be used as a temp storage.It will be automatically cleaned after each run.
+ * @params apiEnvConfig: configs related to api call.
+ */
+object MultiWorkspaceDeployment extends SparkSessionWrapper {
+
+
+  def apply(configCsvPath: String, tempOutputPath: String, apiEnvConfig: ApiEnvConfig): MultiWorkspaceDeployment = {
+    new MultiWorkspaceDeployment()
+      .setConfigCsvPath(configCsvPath)
+      .setOutputPath(tempOutputPath)
+      .setApiEnvConfig(apiEnvConfig)
+      .initDeployment()
+  }
+
+
+  def apply(configCsvPath: String, tempOutputPath: String): MultiWorkspaceDeployment = {
+    new MultiWorkspaceDeployment()
+      .setConfigCsvPath(configCsvPath)
+      .setOutputPath(tempOutputPath)
+      .initDeployment()
+  }
+
+  def apply(configCsvPath: String): MultiWorkspaceDeployment = {
+    new MultiWorkspaceDeployment()
+      .setConfigCsvPath(configCsvPath)
+      .setOutputPath("/mnt/tmp/overwatch")
+      .initDeployment()
+  }
+}
+
+
+class MultiWorkspaceDeployment extends SparkSessionWrapper {
+
+  import spark.implicits._
+
+  private val logger: Logger = Logger.getLogger(this.getClass)
+
+  private var _configCsvPath: String = _
+
+  private var _deploymentId: String = _
+
+  private var _outputPath: String = _
+
+  private var _deploymentReport: ArrayBuffer[MultiWSDeploymentReport] = _
+
+  private var _parallelism: Int = _
+
+  private var _inputDataFrame: DataFrame = _
+
+  private var _apiEnvConfig: ApiEnvConfig = _
+
+  protected def apiEnvConfig: ApiEnvConfig = _apiEnvConfig
+
+  protected def inputDataFrame: DataFrame = _inputDataFrame
+
+  protected def parallelism: Int = _parallelism
+
+  protected def deploymentId: String = _deploymentId
+
+  protected def deploymentReport: ArrayBuffer[MultiWSDeploymentReport] = _deploymentReport
+
+  protected def outputPath: String = _outputPath
+
+  protected def configCsvPath: String = _configCsvPath
+
+  private var _pipelineSnapTime: Long = _
+
```

Comment:
please watch a video on how companion objects work in scala and how to use it to initialize a class. This is not a proper pattern. We can discuss more later.

Reply:
tried to implement and done the code change